server/src/main/java/org/elasticsearch/action/admin/cluster/stats/CCSUsageTelemetry.java

```diff
@@ -0,0 +1,192 @@
+ * can be obtained by getting CCSTelemetrySnapshot objects (TODO: create that class (and determine the best name)).
+ */
+public class CCSUsageTelemetry {
+
+    private final LongAdder totalCCSCount;  // TODO: need this? or just sum the successfulSearchTelem and failedSearchTelem ?
+    private SuccessfulCCSTelemetry successfulSearchTelem;
+    private FailedCCSTelemetry failedSearchTelem;
+    private Map<String, PerClusterCCSTelemetry> byRemoteCluster;
+
+    public CCSUsageTelemetry() {
+        this.totalCCSCount = new LongAdder();
+        this.successfulSearchTelem = new SuccessfulCCSTelemetry();
+        this.failedSearchTelem = new FailedCCSTelemetry();
+        this.byRemoteCluster = new ConcurrentHashMap<>();
+    }
+
+    public void updateUsage(CCSUsage ccsUsage) {
+        // TODO: fork this to a background thread? if yes, could just pass in the SearchResponse to parse it off the response thread
+        doUpdate(ccsUsage);
+    }
+
+    // TODO: what is the best thread-safety model here? Start with locking model in order to get the functionality working.
+    private synchronized void doUpdate(CCSUsage ccsUsage) {
+        totalCCSCount.increment();
+        if (ccsUsage.getFailureType() == null) {
+            // handle successful (or partially successful query)
+            successfulSearchTelem.update(ccsUsage);
+            for (Map.Entry<String, CCSUsage.PerClusterUsage> entry : ccsUsage.getPerClusterUsage().entrySet()) {
+                PerClusterCCSTelemetry perClusterCCSTelemetry = byRemoteCluster.get(entry.getKey());
+                if (perClusterCCSTelemetry == null) {
+                    perClusterCCSTelemetry = new PerClusterCCSTelemetry(entry.getKey());
+                    byRemoteCluster.put(entry.getKey(), perClusterCCSTelemetry);
+                }
+                // TODO: add more fields/data to the perClusterCCSTelemetry
+                perClusterCCSTelemetry.update(entry.getValue());
+            }
+
+        } else {
+            // handle failed query
+            failedSearchTelem.update(ccsUsage);
+        }
+    }
+
+    public long getTotalCCSCount() {
+        return totalCCSCount.sum();
+    }
+
+    // TODO: the getters below need to create an immutable snapshot of CCS Telemetry and return that - see SearchUsageStats as example
+    public SuccessfulCCSTelemetry getSuccessfulSearchTelemetry() {
+        return successfulSearchTelem;
+    }
+
+    public FailedCCSTelemetry getFailedSearchTelemetry() {
+        return failedSearchTelem;
+    }
+
+    public Map<String, PerClusterCCSTelemetry> getTelemetryByCluster() {
+        return byRemoteCluster;
+    }
+
+    /**
+     * Telemetry metrics for successful searches (includes searches with partial failures)
+     */
+    static class SuccessfulCCSTelemetry {
+        private long count; // total number of searches
+        private long countMinimizeRoundtrips;
```

Comment:
Wouldn't we want to know how many searches we did with MRT on or off regardless of whether they were successful? I feel like we should be collecting "what kind of search it is" metrics regardless of how the search ended, since those metrics are established at the start of the search.

Reply:
Or we could just have a single class with failureCauses member? Why we even need a separate class for it?